### DISCUSSION TITLE
How to set a notification sound?

### DISCUSSION THREAD

**Discussion by ghost** [vladmandic/sdnext#1192] [Q&A] [answered]
I've been struggling to get a sounds to play when using a custom mp3 file placed in the root "html" folder where the original notification.mp3 file.

 Any ideas?

**Comment by brknsoul** [comment]
Go to Settings > Live Preview, you can point to a custom notification file there.
Any changes in `html` will be overwritten by an `--upgrade`. It's best to just point to a file elsewhere.

**Reply by ghost** [reply]
I've done that, using the file explorer path but it still wont play, not even the original file. Is there a format that the path needs to be in? Other than **C:/file_folder/notifcation.mp3** _(example)_

**Reply by brknsoul** [reply]
I'm not sure if it makes a difference, but Windows uses black slashes, not forward slashes;
`C:\folder\notification.mp3`

Also, changing the mp3 file won't have effect until a server restart.
If the file doesn't exist, then nothing will be played.

**Reply by vladmandic** [reply]
it has to be in the root of your installation as notification file is downloaded by browser and for security reasons browser is not allowed to access files outside of application root.

**Reply by brknsoul** [reply]
Not quite true, @vladmandic.
My install is located at `G:\automatic`, but I've pointed my "Path to notification sound" setting to `G:\SDfolders\customnotification.mp3`.

`SDfolders` isn't used as any --data-dir settings, it's just where I keep things to backup and symlink (models, repositories, etc)
Perhaps a security oversight? ;-)

**Reply by vladmandic** [reply]
ok, i over-simplified as enforcement is per os and per browser. but best practise is that everything is relative to data-dir (which defaults to application home).

**Comment by ghost** [comment]
> ok, i over-simplified as enforcement is per os and per browser. but best practise is that everything is relative to data-dir (which defaults to application home).

That still doesn't really answer my question... Is there a format that _the path_ needs to be in? Other than **C:/file_folder/notifcation.mp3** (example)

how is it written normally? Does it include the "C:/" or is it a special argument for the path directory?

**Reply by vladmandic** [reply]
REALATIVE TO DATA-DIR. meaning you cannot use absolute paths. 
reason? security enforcements or else web app could fetch anything from your harddrive.

**Reply by brknsoul** [reply]
> how is it written normally?

if you have some file "somefile.mp3" located in the root `automatic` folder, then you'd simply write `somefile.mp3` in the setting.

**Reply by ghost** [reply]
so there is no need to write a the file path? O_o I am going to assume that the **_`automatic`_** folder you are referring to is the `html ` folder where the original `notifcation.mp3` is located. 

That would actually simplify things and would make sense since it is locked to a root folder. I'll try it out and report back...

Update: Tested, still does not work.

**Reply by ghost** [reply]
![info](https://github.com/vladmandic/automatic/assets/59182752/13ca6cbf-db00-4995-bce0-f69c89920d99)

This is what I have set, however there is no sound coming from the program. I am using an mp3 file named blip_blip.mp3 as shown in the image here.

**Reply by brknsoul** [reply]
Yes, and you've put in `automatic\html` instead of the root folder of `automatic`, as per my last comment.
Files in `automatic\html` will get blasted when you run your next `--upgrade`.

>  I am going to assume that the automatic folder you are referring to is the html 

I didn't say _anything_ about `html`. Please try to follow instructions and not assume.

**Reply by ghost** [reply]
That's very rude of you to get onto someone for asking questions about something they need help with. I am also not asking for instructions, just asking for help as to why it isn't working. 

I am very well aware of the overwrite issue with any software updates. Since I am not doing any `--upgrade` commands, that makes your statement irrelevant.
Your 'instructions' of

>if you have some file "somefile.mp3" located in the root automatic folder, then you'd simply write somefile.mp3 in the setting.

was _**NOT**_ addressing the issue I was asking about, nor did you read my question before making that statement. 

**(My original question)**
>I've been struggling to get a sounds to play when using a custom mp3 file placed in the root "html" folder where the original notification.mp3 file.

I am working with the **root** `html` folder since that is the _default_ .mp3 file location. Also it is best to not include changes to the file structure that are unnecessary when troubleshooting something. Placing the file in another folder directory would add an unnecessary change when I want to find the source of the problem first. Once the problem is resolved then we can add other changes that may or may not influence the troubleshooting process. This is something I do as an engineer when troubleshooting for work, and the principle applies here as well. 

**The issue I am having is with the file _not being played_ either with the default file that comes with the software with no modifications, or with the file I want to use.**

My guess of what could be causing the issue, was the way the file-path was being written in the input field. However, with any file-path rewrite does not seem to make it work. 

Unless you have it working yourself, and you have a picture of the filename path in your config options to share, don't post rude comments, or insult my intelligence. 

This is something I hope @vladmandic could better answer since they are the developer of the app, or someone that is willing to share a screenshot of their settings so I can have a visual of their **_working_** settings. 
Keyword it _needs_ to be **_working_** for them already. Otherwise you are just putting in your useless two-cents.

**Reply by vladmandic** [reply]
@NinjaDanny87 slow down - @brknsoul has done nothing but go above and beyond trying to help you!

> Also it is best to not include changes to the file structure that are unnecessary when troubleshooting something

Correct - **unless** explicitly instructed otherwise - which you have. 
Yes, you have right understanding why something should or should not be in `html`, but is that really worth this discussion - why not place it in app home folder and be done with it?

And yes, it does work. But keep in mind few things:
- notifications are triggered only when browser is NOT in focus. I see no point of having notifications triggered always.
- audio notifications are often blocked by browsers. so check your browser.
- change to audio notifications require server restart. why? because audio control is created on server and only `play()` is triggered in js.  
  yes, it could be done on-the-fly, i just don't consider it important enough and normally you set it once and leave it as that.  

Regarding path - its always relative to home folder. So yes, you can put mp3 in a any subfolder or in root or and just reference it like that. There is no special magic.

**Reply by Tillerz** [reply]
Afaik notification sounds are only played when the tab with a1111 is not in front? Might be wrong, but that is my experience. So if you try around with A1111 in front you probably won't here a thing.

**Reply by vladmandic** [reply]
yup - that's what i wrote as the first bullet - notifications are played only when browser is not in focus.

**Reply by ghost** [reply]
Hey @vladmandic,

I am sorry, I wasn't upset nor was it my intention to start an argument. 
I really do appreciate the help @brknsoul has been providing.. however, I won't stand for someone disrespecting me by treating me like a child telling me things like,
> 'follow instructions and not assume.'

I am not a child, nor will I accept to be treated as one.

Again I do apologize if I came off rude myself. 
With that I wish to just move on and let it go. So if everyone is good and chill now, I am good too.

Anyway, that aside,


You mentioned,
> why not place it in app home folder and be done with it?

Even if may not make a difference, it is still software where the code is constantly being worked on. Issues pop up that aren't always apparent until it is tested later, as I am sure you are very well aware.

While troubleshooting, I want to remove as many variables as possible. That is why even though I _was_ instructed to move the file, I kept it in the original directory where the original file was located. It didn't make sense to move it since it was clearly stated that the file would work in either the `html ` folder or the `root ` folder. Why make a change when the original issue was not addressed? I was wanting to fix the issue of the notification sound not playing, and what the correct file path needed to be.  **Not** bypass the file being overwritten.


I did take a look at the points you mentioned, _(which are very good points btw)_ and did the following to test the issue:

- Set browser to not in focus. (this included minimizing, and  having something else full screen during the render process.
And I agree having notify you always is a little annoying so I appreciate that feature.
- I have tried this with multiple browsers (Opera GX, Chrome, Edge) and still the same issue. I have even checked the permissions on each and they are correctly set to allow sound control.
- For each change made to the software's notification settings I have done a full restart. (closing the software completely including the terminal)

I am wondering though, how is the _**file path**_ for the notification file _within_ **VladAutomatic's** _settings_ written normally? Does it include the `"C:/"` or is it like how I have it written in the screenshot I posted earlier? _(where it is just the `notifcation.mp3` file)_

**Reply by vladmandic** [reply, marked as answer]
its written as whatever is the default setting - which happens to be `html/notification.mp3`.
how is the path handled internally is a long, complicated and irrelevant story. what is sent to browser is path relative to data_dir which by default is application home folder. and if you want to investigate, open browser console and find that network request as browser tries to download the mp3 at the app startup and only triggers `play()` method for actual notification.

**Reply by ghost** [reply]
Okay _now_ my question has been answered! Thank you!
> which happens to be `html/notification.mp3`

I just needed the file path as it was inputed into the settings. Now it works! 

My original file path didn't have the `html/` before the `notification.mp3`

I don't know why it was so hard to just answer my question about the file path and how it was written. Lol 😂 

Thank you for your help @vladmandic

**Reply by vladmandic** [reply]
Maybe because that's a default value as seen in the UI and you can always restore to default setting to get it back...

**Reply by ghost** [reply]
For some reason it wasn't like that by default and was blank. Not sure why that was the case. 

It took some figuring out where the file was before I came here. Would there have been a reason for why it wasn't set to default?

I've also noticed alot of my extensions get disabled for no reason even if there wasn't an update.

**Comment by ryonaya** [comment]


**Reply by vladmandic** [reply]
that is exactly what you should NOT do.

**Reply by ryonaya** [reply]
Thank you for replying. 
Just realized it will be replaced by update, sorry for my dumb suggestion, I will remove my answer.
